[mlir] Fix correct memset range in `OwningMemRef` zero-init

chokobole opened the pull request:
### Problem: `OwningMemRef` zero-init can write past allocation

`OwningMemref` allocates with overprovision + manual alignment:

```cpp
auto desiredAlignment = alignment.value_or(nextPowerOf2(sizeof(T))); 
void* data = malloc(size + desiredAlignment); // size = nElements * sizeof(T)

// Compute aligned view
auto addr = reinterpret_cast<uintptr_t>(data);
auto rem  = addr % desiredAlignment;
T* alignedData = (rem == 0)
                   ? static_cast<T*>(data)
                   : reinterpret_cast<T*>(addr + (desiredAlignment - rem));
```

The descriptor stores:

* `descriptor.basePtr = data` (the raw allocation)
* `descriptor.data    = alignedData` (the aligned start)

When creating an `OwningMemRef` **without init**, we zero it. The buggy code does:

```cpp
memset(descriptor.data, 0, size + desiredAlignment); // ❌ may overrun
```

This is invalid because `descriptor.data` (the aligned pointer) **does not point to the full allocated block** (`size + desiredAlignment`). Zeroing that much from the aligned start can write past the end of the allocation.

### Correct zero-init options

Choose **one** of the following:

1. Zero the **entire allocated block** from the raw base:

```cpp
memset(descriptor.basePtr, 0,
       nElements * sizeof(T) + alignment.value_or(nextPowerOf2(sizeof(T))));
```

2. Zero **only the logical payload** (what the user sees) from the aligned start:

```cpp
memset(descriptor.data, 0, nElements * sizeof(T));
```

### Recommendation

I think one of the main root causes here is naming. At first, I assumed `descriptor.`data referred to the raw data pointer—but it doesn’t. Using more consistent naming could make the meaning clearer. For example: `descriptor.data` vs. `descriptor.alignedData`.

joker-eph reviewed (changes requested):
Is this an AI that detected this and wrote this PR? The way the description is phrased looks kind of verbose in the way AI agents tends to be.

Can you add a test please?

chokobole committed:
[mlir] Fix correct memset range in OwningMemRef zero-init

`OwningMemRef` previously called `memset()` on `descriptor.data` with
`size + desiredAlignment`, which could write past the allocated region
since `data != alignedData`.

chokobole commented on mlir/unittests/ExecutionEngine/Invoke.cpp:
```diff
@@ -205,50 +205,66 @@ TEST(NativeMemRefJit, SKIP_WITHOUT_JIT(BasicMemref)) {
-    for (int j = 0; j < m; ++j) {
-      EXPECT_EQ((a[{i, j}]), i * m + j);
-      EXPECT_EQ(&(a[{i, j}]), &((*a)[i][j]));
+  auto runTest = [&](std::string_view traceName,
```
I separated the test and limited to just check the 0 init only!
